src/main/java/apoc/util/s3/S3OutputStream.java

```diff
@@ -0,0 +1,284 @@
+    }
+
+    @Override
+    public void write(@Nonnull final byte[] b) throws IOException {
+        write(b, 0, b.length);
+    }
+
+    // This function call is used directly by OutputStream writer, so it's best that everything routes to it.
+    @Override
+    public void write(@Nonnull final byte[] b, final int offset, final int length) throws IOException {
+        int rdPtr = offset;
+        do {
+            // If the amount of data left to consume from the input is less than the amount of space in the
+            // buffer, fill the remaining space of the buffer with the input, otherwise fully consume the
+            // remaining input.
+            final int wrAmount = Math.min(buffer.length - buffSize, length - (rdPtr - offset));
+            System.arraycopy(b, rdPtr, buffer, buffSize, wrAmount);
+            buffSize += wrAmount;
+            rdPtr += wrAmount;
+
+            // If the buffer is full, transmit it
+            if (buffer.length == buffSize) {
+                transmitBuffer();
+            }
+        } while ((rdPtr - offset) < length);
+    }
+
+    @Override
+    public void close() {
+        // Do not reorder operations, setting done to true first defeats a race condition with the queue being blocked.
+        isDone = true;
+
+        // Based on the requirements of multipart upload, the last piece can disobey the sizing requirements
+        // See https://docs.aws.amazon.com/AmazonS3/latest/dev/qfacts.html
+        queue.add(new S3UploadData(new ByteArrayInputStream(buffer), true, buffSize));
+        buffer = null;
+        try {
+            // Wait for manager's future to complete before exiting.
+            synchronized (managerFuture) {
+                managerFuture.get();
+            }
+        } catch (InterruptedException | ExecutionException e) {
+            if (e instanceof InterruptedException) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    private static class S3UploadData {
+        private final InputStream stream;
+        private final boolean isLast;
+        private final int size;
+
+        S3UploadData(@Nonnull InputStream stream, final boolean isLast, final int size) {
+            this.stream = stream;
+            this.isLast = isLast;
+            this.size = size;
+        }
+
+        InputStream getStream() {
+            return stream;
+        }
+
+        boolean getIsLast() {
+            return isLast;
+        }
+
+        int getSize() {
+            return size;
+        }
+    }
+
+    public class S3UploadManager implements Runnable {
+        private final AmazonS3 s3Client;
+        private final String uploadId;
+        private final BlockingQueue<S3UploadData> queue;
+        private final List<PartETag> partETags = new ArrayList<>();
+        private final InitiateMultipartUploadResult initResponse;
+        private final ExecutorService executorService = Executors.newFixedThreadPool(
```

Comment:
can you please use the job management inside the `Pools` class?
Moreover we cannot use daemon threads please check #1592